### PR TITLE
Fix ModelCheckpoints name formatting

### LINKED ISSUE #2583: Change checkpoint filename without overwriting full filepath
## 🚀 Feature
By default, the checkpoint path resolves to `lightning_logs/{version}/checkpoints/{epoch}.ckpt`. It is currently not possible to change the checkpoint filename without replacing the full path.

### Motivation
The full path is replacable with the `ModelCheckpoint.filepath` argument. This feature is undisirable because I want to use the default path resolver that uses the options from `Trainer` instead of hardcoding a path.

A satisfying path construction workaraound based on values in `Trainer` is not possible because the `Trainer` is constructed *after* the `ModelCheckpoint`.

### Pitch
I would like to add a `filename` parameter to the `ModelCheckpoint` constructor. This parameter is complementary to `filepath` and `filepath` will keep working like it does for backward compatibility.

- If only `filepath` is provided, the original behaviour is used.
- If both `filepath` and `filename` are provided, the checkpoint path will resolve to `filepath`/`filename`.
- If only `filename` is provided, the checkpoint path will resolve to `resolved directory`/`filename` (instead of `resolved directory`/`epoch`).

### LINKED ISSUE #3125: Model checkpoint naming consistency
## 🚀 Feature
1. Modify [`str_ver` inside `format_checkpoin_name()`](https://github.com/PyTorchLightning/pytorch-lightning/blob/18160b81b5c39806b09b3bbdee8b71c316471a43/pytorch_lightning/callbacks/model_checkpoint.py#L267)
2. Let [`ModelCheckpoint.CHECKPOINT_NAME_LAST`](https://github.com/PyTorchLightning/pytorch-lightning/blob/18160b81b5c39806b09b3bbdee8b71c316471a43/pytorch_lightning/callbacks/model_checkpoint.py#L114) contain templates.

### Motivation

1. Going by `format_checkpoint_name`'s [docstring example](https://github.com/PyTorchLightning/pytorch-lightning/blob/18160b81b5c39806b09b3bbdee8b71c316471a43/pytorch_lightning/callbacks/model_checkpoint.py#L245). It is standard to use `-` to split metrics and `_` to split metric names. This makes `str_ver`'s [format](https://github.com/PyTorchLightning/pytorch-lightning/blob/18160b81b5c39806b09b3bbdee8b71c316471a43/pytorch_lightning/callbacks/model_checkpoint.py#L267) inconsistent by using `-`.
2. It would be helpful to be able to set `ModelCheckpoint.CHECKPOINT_NAME_LAST='last-{epoch}-{acc.3f}.ckpt'` with the same parsing behaviour as regular checkpoints.

### Pitch

1. Replace `f"_v{ver}"` for `f"-v{ver}"` or `f".{ver}"`
2. Parse `ModelCheckpoint.CHECKPOINT_NAME_LAST` with `format_checkpoint_name()`

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

- Fixes several typos.
- Adds `CHECKPOINT_JOIN_CHAR`~, `CHECKPOINT_SUFFIX`~ variable~s~ for customization.
- Allows using templates in the last checkpoint saved.
- Adds a test.

Fixes #3125 
Fixes #2583 
Closes #2584 

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

### CODE REVIEW

File: pytorch_lightning/callbacks/model_checkpoint.py
Comment (Borda) [inline pytorch_lightning/callbacks/model_checkpoint.py:116]: maybe rather as it is the same what `os.path.splitext(...)[-1]` returns
```suggestion
    CHECKPOINT_SUFFIX = ".ckpt"
```

**Comment by carmocca** [comment]
Should we add the possibility of modifying these variables to the doc-string?

**Comment by Borda** [comment]
> Should we add the possibility of modifying these variables to the doc-string?

which variables do you mean?

**Comment by carmocca** [comment]
`CHECKPOINT_JOIN_CHAR`, `CHECKPOINT_NAME_LAST`, and `CHECKPOINT_SUFFIX`

**Comment by carmocca** [comment]
I believe this would also solve https://github.com/PyTorchLightning/pytorch-lightning/issues/2583 https://github.com/PyTorchLightning/pytorch-lightning/pull/2584 if merged

File: pytorch_lightning/callbacks/model_checkpoint.py
Comment (rohitgr7) [inline pytorch_lightning/callbacks/model_checkpoint.py:116]: Can we do something like, if `suffix` is present in `filepath`, then keep it as it is else do `.ckpt`. This way we could avoid `CHECKPOINT_SUFFIX`. Also this should be `CHECKPOINT_NAME_LAST=`checkpoint_callback_last_model_path`. What do you think?

**Comment by rohitgr7** [comment]
> I believe this would also solve #2583 #2584 if merged

First, we need to clear up whether we want a separate `filename` parameter or not, because #2583 is a different issue. I don't think this PR solves that problem.

**Comment by carmocca** [comment]
> First, we need to clear up whether we want a separate `filename` parameter or not, because #2583 is a different issue. I don't think this PR solves that problem.

With the changes in this PR, one can do:
```python
checkpoint_callback = pl.callbacks.ModelCheckpoint(filepath=my/file/path/name)
# Generates checkpoints called "name.ckpt" in my/file/path
```
But if filepath is a directory:
```python
checkpoint_callback = pl.callbacks.ModelCheckpoint(filepath=my/file/path/)
# Generates checkpoints called "epoch={epoch}.ckpt" in my/file/path
# note: this is the previous default behaviour
```
So there is no need of a `filename` parameter.

File: pytorch_lightning/callbacks/model_checkpoint.py
Comment (carmocca) [reply pytorch_lightning/callbacks/model_checkpoint.py:116]: `CHECKPOINT_SUFFIX` would not be avoided if we implement `get_suffix_from_filepath()` because we would still need a default behaviour. IMO it is not worth it because we probably want all users to use the same suffix (`ckpt`) for consistency. I just extracted `.ckpt` into a variable to have it in one single place.

Do you mean adding `CHECKPOINT_STATE_LAST_PATH=checkpoint_callback_last_model_path`?

**Comment by ydcjeff** [comment]
can we have a test for this `r = ModelCheckpoint._format_checkpoint_name('path/to/ckpt', 3, {}, prefix='test')` as well?

**Comment by carmocca** [comment]
> can we have a test for this `r = ModelCheckpoint._format_checkpoint_name('path/to/ckpt', 3, {}, prefix='test')` as well?

`_format_checkpoint_name`'s first parameter is assumed to be a `filename`. Your example passes a `filepath` so the result would be `r = test-path/to/ckpt`. I can add tests for `format_checkpoint_name`(no leading _) to test passing filepaths.

File: pytorch_lightning/callbacks/model_checkpoint.py
Comment (awaelchli) [inline pytorch_lightning/callbacks/model_checkpoint.py:408]: can we (in a new PR pls) first rename the model to "oldname.ckpt.backup" before saving the model, and then delete the backup file after save has completed? Otherwise, if a crash happens during save_model, we will have lost the old checkpoint and have a corrupted file for the current.

Comment (carmocca) [reply pytorch_lightning/callbacks/model_checkpoint.py:408]: A new PR based on master or this branch?

Comment (awaelchli) [reply pytorch_lightning/callbacks/model_checkpoint.py:408]: yes, just a small suggestion after this PR is merged. but it will touch other locations as well, for example topk saving.

**Comment by awaelchli** [comment]
I prefer this PR over the one that adds an extra filename parameter.

**Comment by rohitgr7** [comment]
> So there is no need of a filename parameter.

The reason I suggested separate `dirpath` and `filename` parameter is because it gives the functionality to get the default versioning of checkpoints out-of-the-box as if `dirpath` is set to None and `filename` is defined then ModelCheckpoint will use the versioning functionality of checkpoints in `lightning_logs/version_num`. Also, resolve some issues that are handled by his PR and is a bit easy to use and manipulate checkpoints for users.

**Comment by ananthsub** [comment]
> checkpoint_callback = pl.callbacks.ModelCheckpoint(filepath="my/file/path/name")
> # Generates checkpoints called "name.ckpt" in my/file/path

What about the case where the user expects the callback to save checkpoints to directory `my/file/path/name` with name `{epoch}.ckpt`? I think this is another reasonable expectation and is hard to distinguish with `filepath` being overloaded. It's also confusing because you could pass an uncreated directory path as `filepath` to the checkpoint and expect Lightning to make the directory to save checkpoints to. And parsing this is impossible because the file info isn't available yet

**Comment by carmocca** [comment]
I am not against the idea of splitting filepath into dirpath and filename, however, that discussion is out of scope of this PR.

Considering that this PR does not change the API, we could merge this and then (possibly?) work on the filepath improvement.

File: pytorch_lightning/callbacks/model_checkpoint.py
Comment (rohitgr7) [reply pytorch_lightning/callbacks/model_checkpoint.py:116]: Then I don't think it requires a variable here just to specify the default suffix.

Comment (carmocca) [reply pytorch_lightning/callbacks/model_checkpoint.py:116]: Okay then, I will leave the value hard-coded.

**Comment by carmocca** [comment]
Anything holding this back?

There will be some conflicts with #3320. Should be easy to fix

**Comment by Borda** [comment]
Let's process it...

**Comment by Borda** [comment]
> Shall we fix this typo in this PR too?

yes, pls

**Comment by carmocca** [comment]
@ydcjeff @Borda done!

File: pytorch_lightning/callbacks/early_stopping.py
Comment (Borda) [inline pytorch_lightning/callbacks/early_stopping.py:111]:
```suggestion
        error_msg = (f'Early stopping conditioned on metric `{self.monitor}`'
                     f' which is not available. Either add `{self.monitor}` to the return of'
                     ' `validation_epoch_end` or modify your `EarlyStopping` callback to use any of the'
                     f' following: `{"`, `".join(list(logs.keys()))}`')
```

File: pytorch_lightning/callbacks/early_stopping.py
Comment (Borda) [inline pytorch_lightning/callbacks/early_stopping.py:185]:
```suggestion
            m = (f"When using `EvalResult(early_stop_on=X)` or `TrainResult(early_stop_on=X)`"
                 " the 'monitor' key of `EarlyStopping` has no effect. "
                 f" Remove `EarlyStopping(monitor='{self.monitor}')` to fix")
```

File: pytorch_lightning/callbacks/model_checkpoint.py
Comment (Borda) [inline pytorch_lightning/callbacks/model_checkpoint.py:318]:
```suggestion
            rank_zero_warn(f"When using EvalResult(checkpoint_on=X) or TrainResult(checkpoint_on=X)"
                 " the 'monitor' key of ModelCheckpoint has no effect."
                 f" Remove ModelCheckpoint(monitor='{self.monitor}') to fix")
```

File: pytorch_lightning/callbacks/model_checkpoint.py
Comment (Borda) [inline pytorch_lightning/callbacks/model_checkpoint.py:421]:
```suggestion
                f'Epoch {epoch:d}: {self.monitor} reached'
                f' {current:0.5f} (best {self.best_model_score:0.5f}),'
                f' saving model to {filepath} as top {self.save_top_k}')
```

File: pytorch_lightning/callbacks/model_checkpoint.py
Comment (awaelchli) [inline pytorch_lightning/callbacks/model_checkpoint.py:136]: is this a common assumption? I think it would be safer to treat "" the same as None.

File: tests/callbacks/test_model_checkpoint.py
Comment (awaelchli) [inline tests/callbacks/test_model_checkpoint.py:103]: can we have better var names than "r" or "fp"?

Comment (awaelchli) [reply tests/callbacks/test_model_checkpoint.py:103]: also, how will the checkpoint be formatted with save_last=True?

File: pytorch_lightning/callbacks/model_checkpoint.py
Comment (awaelchli) [inline pytorch_lightning/callbacks/model_checkpoint.py:388]: what will the last be formated?
The good thing about having it just named "last.ckpt" was that you could easily load programmatically the last checkpoint without knowing the epoch. It would be great if we could keep that.

**Comment by awaelchli** [comment]
@carmocca does it also fix #2042 ?

File: pytorch_lightning/callbacks/model_checkpoint.py
Comment (carmocca) [reply pytorch_lightning/callbacks/model_checkpoint.py:136]: It was the assumption taken before #3320. I added this to fix the tests and mimic the previous behaviour. I don't mind undoing this change and updating the test instead. However, it might break the workflow of whoever was using `filepath=""`

File: tests/callbacks/test_model_checkpoint.py
Comment (carmocca) [reply tests/callbacks/test_model_checkpoint.py:103]: Updated var names.

What do you mean? There is no `save_last` parameter in the `_format_checkpoint_name` function

File: pytorch_lightning/callbacks/model_checkpoint.py
Comment (carmocca) [reply pytorch_lightning/callbacks/model_checkpoint.py:388]: To allow having the epoch (and any other metrics) in the last checkpoint name is one of the main features of this PR. If the user has not modified `CHECKPOINT_NAME_LAST` the last will just be named `last.ckp`. Just as before.

**Comment by carmocca** [comment]
> @carmocca does it also fix #2042 ?

If i understood that issue correctly, its problem is not being able to use metrics in the checkpoint name (already supported before this PR) but being able to also keep the default Trainer path structure.

This PR does not alter any of the current `dirpath` logic, so it does not fix that issue.

File: pytorch_lightning/callbacks/model_checkpoint.py
Comment (awaelchli) [reply pytorch_lightning/callbacks/model_checkpoint.py:388]: very nice!

File: pytorch_lightning/callbacks/model_checkpoint.py
Comment (awaelchli) [reply pytorch_lightning/callbacks/model_checkpoint.py:136]: I would prefer the other way, but it's probably best to satisfy the tests.

**Comment by carmocca** [comment]
I would like to land this before next release. Any issues, please tell me :smiley:

File: pytorch_lightning/callbacks/model_checkpoint.py
Comment (rohitgr7) [inline pytorch_lightning/callbacks/model_checkpoint.py:392]:
```suggestion
            self._save_model(filepath, trainer, pl_module)
            if self.last_model_path and self.last_model_path != filepath:
                self._del_model(self.last_model_path)
```
I think it should save first and then delete the old one.

File: pytorch_lightning/callbacks/model_checkpoint.py
Comment (rohitgr7) [inline pytorch_lightning/callbacks/model_checkpoint.py:316]:
```suggestion
            rank_zero_warn(
                f"When using `EvalResult(checkpoint_on=X)` or `TrainResult(checkpoint_on=X)`"
                " the 'monitor' key of `ModelCheckpoint` has no effect."
                f" Remove `ModelCheckpoint(monitor='{self.monitor}')` to fix."
            )
```

File: pytorch_lightning/callbacks/early_stopping.py
Comment (rohitgr7) [inline pytorch_lightning/callbacks/early_stopping.py:186]:
```suggestion
            rank_zero_warn(
                f"When using `EvalResult(early_stop_on=X)` or `TrainResult(early_stop_on=X)`"
                " the 'monitor' key of `EarlyStopping` has no effect. "
                f" Remove `EarlyStopping(monitor='{self.monitor}')` to fix.")
            )
```

File: pytorch_lightning/callbacks/model_checkpoint.py
Comment (rohitgr7) [inline pytorch_lightning/callbacks/model_checkpoint.py:136]:
```suggestion
        if not filepath:  # will be determined by trainer at runtime
            self.dirpath, self.filename = None, None
```
should we do this?

File: pytorch_lightning/callbacks/model_checkpoint.py
Comment (rohitgr7) [inline pytorch_lightning/callbacks/model_checkpoint.py:239]:
```suggestion
        if not filename:
            # filename is not set, use default name
            filename = '{epoch}'
        
        # check and parse user passed keys in the string
        groups = re.findall(r'(\{.*?)[:\}]', filename)
        if len(groups):
            metrics['epoch'] = epoch
            for tmp in groups:
                name = tmp[1:]
                filename = filename.replace(tmp, name + '={' + name)
                if name not in metrics:
                    metrics[name] = 0
            filename = filename.format(**metrics)
```
let regex do the formatting.

File: pytorch_lightning/callbacks/model_checkpoint.py
Comment (rohitgr7) [inline pytorch_lightning/callbacks/model_checkpoint.py:264]:
```suggestion
        if ver is not None:
            filename = self.CHECKPOINT_JOIN_CHAR.join([filename, f'v{ver}'])
        
        return os.path.join(self.dirpath, f'{filename}.ckpt')
```

File: pytorch_lightning/callbacks/model_checkpoint.py
Comment (carmocca) [reply pytorch_lightning/callbacks/model_checkpoint.py:392]: That does not work when you have a checkpoint from a previous run with the same name.

Say you already have a last checkpoint from a previous run named "last.ckpt" (default name). When you run your experiment again, your change would overwrite the previous "last.ckpt" and then delete it, losing both checkpoints in the process.

File: pytorch_lightning/callbacks/model_checkpoint.py
Comment (carmocca) [reply pytorch_lightning/callbacks/model_checkpoint.py:136]: Discussed here: https://github.com/PyTorchLightning/pytorch-lightning/pull/3163#discussion_r483197256

Comment (rohitgr7) [reply pytorch_lightning/callbacks/model_checkpoint.py:136]: `filepath=""` is just an empty string equivalent to None. I prefer `"" the same as None, and have "." and "./" lead to cwd` as @awaelchli said and maybe update the test if they are failing.

File: pytorch_lightning/callbacks/model_checkpoint.py
Comment (carmocca) [reply pytorch_lightning/callbacks/model_checkpoint.py:393]: You are correct. Changed!

File: pytorch_lightning/callbacks/model_checkpoint.py
Comment (carmocca) [reply pytorch_lightning/callbacks/model_checkpoint.py:136]: Done.

**Review by SkafteNicki — APPROVED** [review]

File: pytorch_lightning/callbacks/model_checkpoint.py
Comment (rohitgr7) [inline pytorch_lightning/callbacks/model_checkpoint.py:262]: here `self.dirpath` can't be None right? if it's none in init it will be assigned to `default_root_dir` from Trainer.

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/callbacks/model_checkpoint.py
Comment (carmocca) [reply pytorch_lightning/callbacks/model_checkpoint.py:262]: Should not be None during normal usage, however, it is in this case (covered by a test):
https://github.com/PyTorchLightning/pytorch-lightning/blob/2d17626f902cd7822b3d8539acb919a240203ffd/tests/callbacks/test_model_checkpoint.py#L135

File: tests/callbacks/test_model_checkpoint.py
Comment (Borda) [inline tests/callbacks/test_model_checkpoint.py:130]:
```suggestion
    char_org = ModelCheckpoint.CHECKPOINT_JOIN_CHAR
    ModelCheckpoint.CHECKPOINT_JOIN_CHAR = '@'
    ckpt_name = ModelCheckpoint._format_checkpoint_name('{epoch},{acc:.5f}', 3, {'acc': 0.03}, prefix='test')
    assert ckpt_name == 'test@epoch=3,acc=0.03000'
    ModelCheckpoint.CHECKPOINT_JOIN_CHAR = char_org
```

File: tests/callbacks/test_model_checkpoint.py
Comment (Borda) [inline tests/callbacks/test_model_checkpoint.py:114]:
```suggestion
def test_model_checkpoint_format_checkpoint_name(tmpdir):
```

File: tests/callbacks/test_model_checkpoint.py
Comment (Borda) [inline tests/callbacks/test_model_checkpoint.py:133]:
```suggestion
```

File: tests/trainer/test_trainer.py
Comment (Borda) [inline tests/trainer/test_trainer.py:418]:
```suggestion
        f" Should save {len(expected_files)} models when save_top_k={save_top_k} but found={file_lists}"
```

File: tests/trainer/test_trainer.py
Comment (Borda) [inline tests/trainer/test_trainer.py:417]:
```suggestion
        f"Should save {len(expected_files)} models when save_top_k={save_top_k} but found={file_lists}"
```

**Review by Borda — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]
I think this is a great improvement. Thanks for your effort. 
Please note for the future, this PR has some independent changes that, although they are good to have, they could have been split over multiple PR's.

**Comment by awaelchli** [comment]
> Parsing whether the last component should be part of the directory or the checkpoint name is impossible to do right, and suggests better support from the API arguments. IMO splitting filepath into 2 args, dirpath and file_template_name (with default value = "{epoch}"), would resolve this nicely.

as [mentioned](https://github.com/PyTorchLightning/pytorch-lightning/pull/3163#issuecomment-683539565) by @ananthsub it is not possible to distinguish between filename and folder, this bug: #3001
My suggestion was to force user to choose a file extension so that we know when it is a file. But perhaps a separate arg is nicer. 
Also there is a 2nd bug which complicates things, self._fs.isdir checks for a real existing dir, but we want  to accept a folder name that does not exist yet.

Either way, to fix this I don't see a good way that keeps back compatibility. If back compatibility is unavoidable, I feel like now with the v1 release would be a good time to introduce this breaking change. If not, we must at least update docs. 
@PyTorchLightning/core-contributors

**Comment by carmocca** [comment]
> Either way, to fix this I don't see a good way that keeps back compatibility.

Cant we keep back compat by having filepath (old parameter) and dirpath, filename (new parameters)?

- if filepath is set: old logic, show deprecated warning.
- if dirpath and filename are set: new logic.
- if filepath, dirpath, and filename are set: raise MisconfigurationException or use new logic

**Comment by Borda** [comment]
> Also there is a 2nd bug which complicates things, self._fs.isdir checks for a real existing dir, but we want to accept a folder name that does not exist yet.

then we can do some bubble up testing to find the existing folder in upstream and create the remaining?

> Either way, to fix this I don't see a good way that keeps back compatibility. If back compatibility is unavoidable, I feel like now with the v1 release would be a good time to introduce this breaking change. If not, I think we should at least update docs.

yes, 1.0 is a good place to do it...

**Comment by Borda** [comment]
> BTW this discussion should continue elsewhere. Maybe in the issue you linked.

@carmocca mind creating a new issue for this discussion and link it here...

**Pull request merged**
